test/sql-features/mzcompose.py

```diff
@@ -0,0 +1,219 @@
+
+    @classmethod
+    def phase2(cls) -> str:
+        return "\n\n".join(
+            [
+                # Include the header.
+                header(f"{cls.__name__} (phase 2)", drop_schema=False),
+                # We can query item #1 when the feature is turned on. Ensures
+                # that catalog rehydration ignores SQL-level feature flags.
+                query_ok(cls.query_item(ordinal=1)),
+                # We can drop item #1.
+                statement_ok(cls.drop_item(ordinal=1)),
+                # We cannot create item #2 when the feature is turned off.
+                # Ensures that the feature flag is respected for new items.
+                statement_error(cls.create_item(ordinal=2), cls.feature_error()),
+            ]
+        )
+
+    @classmethod
+    def feature_name(cls) -> str:
+        """The name of the feature flag under test."""
+        assert False, "feature_name() must be overriden"
+
+    @classmethod
+    def feature_error(cls) -> str:
+        """The error expected when the feature is disabled."""
+        assert False, "feature_error() must be overriden"
+
+    @classmethod
+    def create_item(cls, ordinal: int) -> str:
+        """A SQL statement that creates an item that depends on the feature."""
+        assert False, "create_item() must be overriden"
+
+    @classmethod
+    def drop_item(cls, ordinal: int) -> str:
+        """A SQL statement that drops an item that depends on the feature."""
+        assert False, "drop_item() must be overriden"
+
+    @classmethod
+    def query_item(cls, ordinal: int) -> str:
+        """A SQL query referencing an item that depends on the feature."""
+        assert False, "query_item() must be overriden"
+
+
+class WithMutuallyRecursive(FeatureTestScenario):
+    @classmethod
+    def feature_name(cls) -> str:
+        return "enable_with_mutually_recursive"
+
+    @classmethod
+    def feature_error(cls) -> str:
+        return "`WITH MUTUALLY RECURSIVE` syntax is not enabled"
+
+    @classmethod
+    def create_item(cls, ordinal: int) -> str:
+        return dedent(
+            f"""
+            CREATE VIEW wmr_{ordinal:02d} AS WITH MUTUALLY RECURSIVE
+                foo (a int, b int) AS (SELECT 1, 2 UNION SELECT a, 7 FROM bar),
+                bar (a int) as (SELECT a FROM foo)
+            SELECT * FROM bar;
+            """
+        )
+
+    @classmethod
+    def drop_item(cls, ordinal: int) -> str:
+        return f"DROP VIEW wmr_{ordinal:02d}"
+
+    @classmethod
+    def query_item(cls, ordinal: int) -> str:
+        return f"SELECT * FROM wmr_{ordinal:02d}"
+
+
+def run_test(c: Composition, args: argparse.Namespace) -> None:
+    c.up("materialized")
+    c.up("testdrive", persistent=True)
+
+    scenarios = (
+        [globals()[args.scenario]]
+        if args.scenario
+        else FeatureTestScenario.__subclasses__()
+    )
+
+    for scenario in scenarios:
+        with tempfile.NamedTemporaryFile(mode="w", dir=c.path) as tmp:
```

Comment:
There is no longer a need to create temporary files, you can simply do:

```
c.testdrive(input="your string goes here")
```

My suggestion is to simply pass `c: Composition` to all code that is currently generating testdrive fragments. That code can then use `c.testdrive` to simply run the fragment right away without having to pass it around.

Reply:
The fragments are quite longa and enclosing them in tempfile creates some nice sections in the build:

https://buildkite.com/materialize/nightlies/builds/1912#0186c12d-1188-45f0-bfee-bc913c96bcd9